README.md
# NAME

Amon2::Lite - Sinatra-ish framework on Amon2!

# SYNOPSIS

    use Amon2::Lite;

    get '/' => sub {
        my ($c) = @_;
        return $c->render('index.tt');
    };

    __PACKAGE__->to_app();

    __DATA__

    @@ index.tt
    <!doctype html>
    <html>
        <body>Hello</body>
    </html>

# DESCRIPTION

This is a Sinatra-ish wrapper for Amon2.

__THIS MODULE IS BETA STATE. API MAY CHANGE WITHOUT NOTICE__.

# FUNCTIONS

- `any(\@methods, $path, \&code)`
- `any($path, \&code)`

    Register new route for router.

- `get($path, $code->($c))`

    Register new route for router.

- `post($path, $code->($c))`

    Register new route for router.

- `__PACKAGE__->load_plugin($name, \%opts)`

    Load a plugin to the context object.

- \[EXPERIMENTAL\] `__PACKAGE__->enable_session(%args)`

    This method enables [Plack::Middleware::Session](https://metacpan.org/pod/Plack::Middleware::Session).

    `%args` would be pass to enabled to `Plack::Middleware::Session->new`.

    The default state class is [Plack::Session::State::Cookie](https://metacpan.org/pod/Plack::Session::State::Cookie), and store class is [Plack::Session::Store::File](https://metacpan.org/pod/Plack::Session::Store::File).

    This option enables a response filter, that adds ` Cache-Control: private ` header.

- \[EXPERIMENTAL\] `__PACKAGE__->enable_middleware($klass, %args)`

        __PACKAGE__->enable_middleware('Plack::Middleware::XFramework', framework => 'Amon2::Lite');

    Enable the Plack middlewares.

- `__PACKAGE__->to_app(%args)`

    Create new PSGI application instance.

    There is a options.

    - `no_x_content_type_options : default false`

            __PACKAGE__->to_app(no_x_content_type_options => 1);

        Amon2::Lite puts `X-Content-Type-Options` header by default for security reason.
        You can disable this feature by this option.

    - `no_x_frame_options`

            __PACKAGE__->to_app(no_x_frame_options => 1);

        Amon2::Lite puts `X-Frame-Options: DENY` header by default for security reason.
        You can disable this feature by this option.

# FAQ

- How can I configure the options for Xslate?

    You can provide a constructor arguments by configuration.
    Write following lines on your app.psgi.

        __PACKAGE__->template_options(
            syntax => 'Kolon',
        );

- How can I use other template engines instead of Text::Xslate?

    You can use any template engine with Amon2::Lite. You can overwrite create\_view method same as normal Amon2.

    This is a example to use [Text::MicroTemplate::File](https://metacpan.org/pod/Text::MicroTemplate::File).

        use Tiffany::Text::MicroTemplate::File;

        sub create_view {
            Tiffany::Text::MicroTemplate::File->new(+{
                include_path => ['./tmpl/']
            })
        }

- How can I handle static files?

    If you pass the 'handle\_static' option to 'to\_app' method, Amon2::Lite handles /static/ path to ./static/ directory.

        use Amon2::Lite;
        __PACKAGE__->to_app(handle_static => 1);

- Where is a example codes?

    There is a tiny TinyURL example: [https://github.com/tokuhirom/MyTinyURL/blob/master/app.psgi](https://github.com/tokuhirom/MyTinyURL/blob/master/app.psgi).

- How can I use session?

    You can enable session by `__PACKAGE__->enable_session()`. And you can access the session object by `$c->session` accessor.

        use Amon2::Lite;

        get '/' => sub {
            my $c = shift;
            my $cnt = $c->session->get('cnt') || 1;
            $c->session->set('cnt' => $cnt+1);
            return $c->create_response(200, [], [$cnt]);
        };

        __PACKAGE__->enable_session(); # 
        __PACKAGE__->to_app();

# AUTHOR

Tokuhiro Matsuno <tokuhirom AAJKLFJEF@ GMAIL COM>

# SEE ALSO

# LICENSE

Copyright (C) Tokuhiro Matsuno

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.


Changes
Revision history for Perl extension Amon2::Lite

{{$NEXT}}

0.13 2014-08-28T09:41:58Z

    - Fixed deps.
      (Reported by andk)

0.12 2014-03-25T01:06:43Z

    - Fix default DB name to expand environment name
      (moznion)
    - Fixed testing issues
      (Kosuke Asami)

0.11 2013-10-08T23:29:10Z

    - Amon2::Lite with Text::Xslate 2.0010- gets Perl 5.18+ issue.
      Just updated deps.
      ref. http://blog.64p.org/entry/2013/10/08/093236
      (tokuhirom)
    - Tiny fix for Amon2 5.00 compatibility.
      (tokuhirom)

0.10 2013-08-14T01:14:41Z

    - minil migrate
      (tokuhirom)

0.09

    - Allows "get '/'" and "post '/'" in same router.
      (Reported by mattn++)

0.08

    [bug fix]
    commit 60372b2ac2af6d6a2b0683626e1bf7650e7b5b60
    Author: Syohei YOSHIDA <syohex@gmail.com>
    Date:   Thu Jul 12 00:27:23 2012 +0900

    fix typo

    lib/Amon2/Setup/Flavor/Lite.pm |    2 +-
    1 file changed, 1 insertion(+), 1 deletion(-)

0.07

    [BUG FIX]
    - 'any' was broken at 0.06

0.06

    * e26e2d3 better coding style
    * 936a1e4 better static file handling
    * cbe3fca puts x-nantoka headers by default
    * b857d56 simplified code
    * 8e20caa added enable_session method
    * added enable_middleware method
    * Return 405 Method Not Allowed if it's not allowed(Old impl returns 404)
    * 52a9094 docs about session

0.05

    [SETUP FLAVOR]
    * c4b5db2 ignore .gitignore files
    * d6e0163 better html skelton
    * eac6045 setup bootstrap and jquery by default.

    [DOC ENHANCEMENTS]
    * a27cd94 link to example code

0.04

    - added 'handle_static' option.
    - remove some deps. 

0.03

    - added 'template_options' method.
      (suggested by gfx)

0.02

    - fixed incompatibility with latest Amon2::Lite.
      Amon2::Lite should not be use Bridge::Star.

0.01    Mon Nov 14 09:24:15 2011
    - original version


t/TestFlavor.pm
use strict;
use warnings;
use utf8;

package t::TestFlavor;
use parent qw(Exporter);
our @EXPORT = qw(test_flavor);
use File::Temp qw/tempdir/;
use App::Prove;
use File::Basename;
use Cwd;
use File::Spec;
use Plack::Util;
use Test::More;

sub test_flavor {
    my ($code, $flavor_class) = @_;

	local $ENV{PLACK_ENV} = 'development';
    $flavor_class = Plack::Util::load_class($flavor_class, 'Amon2::Setup::Flavor');

    my $libpath = File::Spec->rel2abs(File::Spec->catfile(dirname(__FILE__), '..', '..', 'lib'));
    unshift @INC, $libpath;

    my $dir = tempdir(CLEANUP => $ENV{DEBUG} ? 0 : 1);
    my $cwd = Cwd::getcwd();
    chdir($dir);
    note $dir;

    {
        my $flavor = $flavor_class->new(module => 'My::App');
        $flavor->run;
        $code->($flavor);

        # run prove
        my $app = App::Prove->new();
        $app->process_args('-Ilib', "-I$libpath", <t/*.t>);
        ok($app->run);
    }

    chdir($cwd);
}

1;



Build.PL
# =========================================================================
# THIS FILE IS AUTOMATICALLY GENERATED BY MINILLA.
# DO NOT EDIT DIRECTLY.
# =========================================================================

use 5.008_001;

use strict;
use warnings;
use utf8;

use Module::Build;
use File::Basename;
use File::Spec;
use CPAN::Meta;
use CPAN::Meta::Prereqs;

my %args = (
    license              => 'perl',
    dynamic_config       => 0,

    configure_requires => {
        'Module::Build' => 0.38,
    },

    name            => 'Amon2-Lite',
    module_name     => 'Amon2::Lite',
    allow_pureperl => 0,

    script_files => [glob('script/*'), glob('bin/*')],
    c_source     => [qw()],
    PL_files => {},

    test_files           => ((-d '.git' || $ENV{RELEASE_TESTING}) && -d 'xt') ? 't/ xt/' : 't/',
    recursive_test_files => 1,

    
);
if (-d 'share') {
    $args{share_dir} = 'share';
}

my $builder = Module::Build->subclass(
    class => 'MyBuilder',
    code => q{
        sub ACTION_distmeta {
            die "Do not run distmeta. Install Minilla and `minil install` instead.\n";
        }
        sub ACTION_installdeps {
            die "Do not run installdeps. Run `cpanm --installdeps .` instead.\n";
        }
    }
)->new(%args);
$builder->create_build_script();

my $mbmeta = CPAN::Meta->load_file('MYMETA.json');
my $meta = CPAN::Meta->load_file('META.json');
my $prereqs_hash = CPAN::Meta::Prereqs->new(
    $meta->prereqs
)->with_merged_prereqs(
    CPAN::Meta::Prereqs->new($mbmeta->prereqs)
)->as_string_hash;
my $mymeta = CPAN::Meta->new(
    {
        %{$meta->as_struct},
        prereqs => $prereqs_hash
    }
);
print "Merging cpanfile prereqs to MYMETA.yml\n";
$mymeta->save('MYMETA.yml', { version => 1.4 });
print "Merging cpanfile prereqs to MYMETA.json\n";
$mymeta->save('MYMETA.json', { version => 2 });


lib/Amon2/Setup/Flavor/Lite.pm
use strict;
use warnings;
use utf8;

package Amon2::Setup::Flavor::Lite;
use parent qw(Amon2::Setup::Flavor);

sub run {
    my ($self) = @_;

    require Amon2::Lite;
    require Amon2;

    $self->{amon2_version}      = $Amon2::VERSION;
    $self->{amon2_lite_version} = $Amon2::Lite::VERSION;

    $self->write_file('app.psgi', <<'...');
use strict;
use warnings;
use utf8;
use File::Spec;
use File::Basename;
use lib File::Spec->catdir(dirname(__FILE__), 'extlib', 'lib', 'perl5');
use lib File::Spec->catdir(dirname(__FILE__), 'lib');
use Amon2::Lite;

our $VERSION = '0.12';

# put your configuration here
sub load_config {
    my $c = shift;

    my $mode = $c->mode_name || 'development';

    +{
        'DBI' => [
            "dbi:SQLite:dbname=$mode.db",
            '',
            '',
        ],
    }
}

get '/' => sub {
    my $c = shift;
    return $c->render('index.tt');
};

# load plugins
__PACKAGE__->load_plugin('Web::CSRFDefender' => {
    post_only => 1,
});
# __PACKAGE__->load_plugin('DBI');
# __PACKAGE__->load_plugin('Web::FillInFormLite');
# __PACKAGE__->load_plugin('Web::JSON');

__PACKAGE__->enable_session();

__PACKAGE__->to_app(handle_static => 1);

__DATA__

@@ index.tt
<!doctype html>
<html>
<head>
    <meta charset="utf-8">
    <title><% $module %></title>
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <script type="text/javascript" src="http://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
    <script type="text/javascript" src="[% uri_for('/static/js/main.js') %]"></script>
    <link href="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.1/css/bootstrap-combined.min.css" rel="stylesheet">
    <script src="//netdna.bootstrapcdn.com/twitter-bootstrap/2.3.1/js/bootstrap.min.js"></script>
    <link rel="stylesheet" href="[% uri_for('/static/css/main.css') %]">
</head>
<body>
    <div class="container">
        <header><h1><% $module %></h1></header>
        <section class="row">
            This is a <% $module %>
        </section>
        <footer>Powered by <a href="http://amon.64p.org/">Amon2::Lite</a></footer>
    </div>
</body>
</html>

@@ /static/js/main.js

@@ /static/css/main.css
footer {
    text-align: right;
}
...

    $self->write_file('Makefile.PL', <<'...');
use ExtUtils::MakeMaker;

WriteMakefile(
    NAME          => '<% $module %>',
    AUTHOR        => 'Some Person <person@example.com>',
    VERSION_FROM  => 'app.psgi',
    PREREQ_PM     => {
        'Amon2'                           => '<% $amon2_version %>',
        'Amon2::Lite'                     => '<% $amon2_lite_version %>',
        'Text::Xslate'                    => '1.5006',
        'Plack::Session'                  => '0.14',
    },
    MIN_PERL_VERSION => '5.008001',
    (-d 'xt' and $ENV{AUTOMATED_TESTING} || $ENV{RELEASE_TESTING}) ? (
        test => {
            TESTS => 't/*.t xt/*.t',
        },
    ) : (),
);
...

    $self->write_file('t/Util.pm', <<'...');
package t::Util;
BEGIN {
    unless ($ENV{PLACK_ENV}) {
        $ENV{PLACK_ENV} = 'test';
    }
}
use parent qw/Exporter/;
use Test::More 0.96;

our @EXPORT = qw//;

{
    # utf8 hack.
    binmode Test::More->builder->$_, ":utf8" for qw/output failure_output todo_output/;
    no warnings 'redefine';
    my $code = \&Test::Builder::child;
    *Test::Builder::child = sub {
        my $builder = $code->(@_);
        binmode $builder->output,         ":utf8";
        binmode $builder->failure_output, ":utf8";
        binmode $builder->todo_output,    ":utf8";
        return $builder;
    };
}

1;
...

    $self->write_file('t/01_root.t', <<'...');
use strict;
use warnings;
use utf8;
use t::Util;
use Plack::Test;
use Plack::Util;
use Test::More;

my $app = Plack::Util::load_psgi 'app.psgi';
test_psgi
    app => $app,
    client => sub {
        my $cb = shift;
        my $req = HTTP::Request->new(GET => 'http://localhost/');
        my $res = $cb->($req);
        is $res->code, 200;
        diag $res->content if $res->code != 200;
    };

done_testing;
...

    $self->write_file('xt/03_pod.t', <<'...');
use Test::More;
eval "use Test::Pod 1.00";
plan skip_all => "Test::Pod 1.00 required for testing POD" if $@;
all_pod_files_ok();
...

    $self->write_file('.gitignore', <<'...');
Makefile
inc/
MANIFEST
*.bak
*.old
nytprof.out
nytprof/
*.db
blib/
pm_to_blib
META.json
META.yml
MYMETA.json
MYMETA.yml
...
}

1;
__END__

=head1 NAME

Amon2::Setup::Flavor::Lite - Amon2::Lite flavor

=head1 SYNOPSIS

    % amon2-setup.pl --flavor=Lite MyApp

=head1 DESCRIPTION

This is a flavor for project using Amon2::Lite.

=head1 AUTHOR

Tokuhiro Matsuno
